x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/action/TransportQuerySearchApplicationAction.java

```diff
@@ -0,0 +1,117 @@
+import org.elasticsearch.script.TemplateScript;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xcontent.NamedXContentRegistry;
+import org.elasticsearch.xcontent.XContentFactory;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+import org.elasticsearch.xcontent.XContentType;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class TransportQuerySearchApplicationAction extends SearchApplicationTransportAction<
+    QuerySearchApplicationAction.Request,
+    SearchResponse> {
+
+    private static final Logger logger = LogManager.getLogger(TransportQuerySearchApplicationAction.class);
+
+    private final Client client;
+    private final ScriptService scriptService;
+
+    private final NamedXContentRegistry xContentRegistry;
+
+    @Inject
+    public TransportQuerySearchApplicationAction(
+        TransportService transportService,
+        ActionFilters actionFilters,
+        Client client,
+        ClusterService clusterService,
+        NamedWriteableRegistry namedWriteableRegistry,
+        NamedXContentRegistry xContentRegistry,
+        BigArrays bigArrays,
+        ScriptService scriptService,
+        XPackLicenseState licenseState
+    ) {
+        super(
+            QuerySearchApplicationAction.NAME,
+            transportService,
+            actionFilters,
+            QuerySearchApplicationAction.Request::new,
+            client,
+            clusterService,
+            namedWriteableRegistry,
+            bigArrays,
+            licenseState
+        );
+        this.client = client;
+        this.scriptService = scriptService;
+        this.xContentRegistry = xContentRegistry;
+    }
+
+    @Override
+    protected void doExecute(QuerySearchApplicationAction.Request request, ActionListener<SearchResponse> listener) {
+        systemIndexService.getSearchApplication(request.name(), listener.delegateFailure((l, searchApplication) -> {
+            final Script script = searchApplication.searchApplicationTemplate().script();
+
+            try {
+                final SearchSourceBuilder sourceBuilder = renderTemplate(script, mergeTemplateParams(request, script));
+                SearchRequest searchRequest = new SearchRequest(searchApplication.indices()).source(sourceBuilder);
+
+                client.execute(
+                    SearchAction.INSTANCE,
+                    searchRequest,
+                    listener.delegateFailure((l2, searchResponse) -> l2.onResponse(searchResponse))
+                );
+            } catch (IOException e) {
+                l.onFailure(e);
+            }
+        }));
+    }
+
+    private static Map<String, Object> mergeTemplateParams(QuerySearchApplicationAction.Request request, Script script) {
+        Map<String, Object> mergedTemplateParams = new HashMap<>(script.getParams());
```

Comment:
I think we should use `XContentHelper#mergeDefaults` here. It recursively merges the inner maps so that partial updates are possible. 

So like adding a new sub map: 

```
{
   "query": {
       "fields": {
             "text": {
                   "boost": 2.4
              }
       }
}
```
and the update is:
```
{
   "query": {
       "fields": {
             "title": {
                   "boost": 2.4
              }
       }
}
```
Results should be:
```
{
   "query": {
       "fields": {
             "text": {
                   "boost": 2.4
              },
              "title": {
                   "boost": 2.4
              }
       }
}
```

Should also works to add a new field:
```
{
   "query": {
       "fields": {
             "text": {
                   "operator": "and"
              }
       }
}
```
results:
```
{
   "query": {
       "fields": {
             "text": {
                   "boost": 2.4,
                   "operator": "and"
              },
              "title": {
                   "boost": 2.4
              }
       }
}
```

Reply:
Thanks @jimczi ! I created https://github.com/elastic/enterprise-search-team/issues/4328 to implement these improvements. 👍